tools/lint/lint.py

```diff
@@ -543,41 +551,52 @@ def check_parsed(repo_root, path, f):
                 if variant != "" and variant[0] not in ("?", "#"):
                     errors.append(rules.MalformedVariant.error(path, (path,)))
 
-        seen_elements = {"timeout": False,
-                         "testharness": False,
-                         "testharnessreport": False}
-        required_elements = [key for key, value in {"testharness": True,
-                                                    "testharnessreport": len(testharnessreport_nodes) > 0,
-                                                    "timeout": len(source_file.timeout_nodes) > 0}.items()
-                             if value]
+        required_elements.extend(key for key, value in {"testharness": True,
+                                                        "testharnessreport": len(testharnessreport_nodes) > 0,
+                                                        "timeout": len(source_file.timeout_nodes) > 0}.items()
+                                 if value)
+
+    testdriver_vendor_nodes = []  # type: List[ElementTree.Element]
+    if source_file.testdriver_nodes:
+        if len(source_file.testdriver_nodes) > 1:
+            errors.append(rules.MultipleTestdriver.error(path))
+
+        testdriver_vendor_nodes = source_file.root.findall(".//{http://www.w3.org/1999/xhtml}script[@src='/resources/testdriver-vendor.js']")
+        if not testdriver_vendor_nodes:
+            errors.append(rules.MissingTestdriverVendor.error(path))
+        else:
+            if len(testdriver_vendor_nodes) > 1:
+                errors.append(rules.MultipleTestdriverVendor.error(path))
+
+        required_elements.extend(["testdriver", "testdriver-vendor"])
```

Comment:
If you don't have a testdriver-vendor node, the inclusion of it in required_elements here will force us to walk every element in `source_file.root.iter()`, because `all(seen_elements[name] for name in required_elements)` will never be true.

I believe that that is the reason why line 555 has `"testharnessreport": len(testharnessreport_nodes) > 0,`, even though not having (exactly) one testharness report node is an error in the file.

Reply:
Yeah that's what the optimization is for.

```suggestion
        required_elements.append("testdriver")
        if len(testdriver_vendor_nodes) > 0:
            required_elements.append("testdriver-vendor")
```